design-docs/commands-v3.md

```diff
@@ -0,0 +1,305 @@
+continuation and moves on to the next command in the list.
+
+Coroutines are essentially wrappers around a continuation primitive that allows additional access to
+the scheduler and provides async/await-like functionality. For example, `Coroutine.await(Command)`
+will schedule a given command and call `yield()` in a loop until that command has completed
+execution (note: this does *not* naively call `command.run(coroutine)`, which would hide the inner
+command from the scheduler and potentially sidestep requirement mutexing)
+
+### Command Function Bodies
+
+Command logic lives in a single function `run` that accepts a `Coroutine` object argument. All
+yielding is done via that coroutine, as is any management of nested commands (scheduling, awaiting,
+cancelling, etc).
+
+The coroutine’s `yield()` method always returns `true`. When a command is cancelled, it is removed
+from the set of running commands and its carrier objects left to be garbage collected. An
+`onCancel()` method on the cancelled command will be invoked to let the command do any necessary
+cleanup; this is needed because the command body will not be re-mounted and run, so a separate
+function is used. Command builders allow this to be specified with a `whenCancelled(Runnable)` step.
+
+### Command Lifetimes
+
+All commands are managed and executed by the scheduler. No commands should manually run their inner
+commands (which is what the old sequential and parallel groups did), since that hides those inner
+commands from the scheduler and telemetry. However, this is not prohibited, nor would there be a way
+to detect it.
+
+Inner commands are bound to the lifetime of their parents. Cancelling a higher-level command must
+also cancel all the commands it scheduled (and so on down the hierarchy). Suspending a higher-level
+command must likewise also suspend all inner commands, and resuming the higher-level command must
+resume all suspended inner commands. Inner commands cannot be resumed before their parent command,
+but may be suspended while the parent command still runs.
+
+When a command is running, the scheduler tracks in a wrapper CommandState object that includes the
+command, the command that scheduled it (if applicable), and the coroutine object carrying the
+command.
+
+Each command run has its own coroutine to back its execution. When a command finishes, its backing
+continuation is done, and the command is removed from the scheduler. Any unfinished inner commands
+are cancelled at this point.
+
+### Factory Methods and Builders
+
+In contrast with v2’s decorator API that returns a new command object for each modification (which
+had some issues with naming and making telemetry difficult with deeply nested wrapper objects), v3
+uses builder objects that allow configuration to be set before creating a single command at the end.
+
+#### Forced Naming
+
+Builder objects will only permit command creation when a name for the command has been provided.
```

Comment:
The current plan is for the terminal build method to accept a string, so by convention command construction would look like this:

```java
Command cmd =
  Command.builder()
    // ... builder configuration
    .named("The Command Name");
```